File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/Attachments/CopyMoveRename.cs
using System.Threading.Tasks;
using Raven.Client.Documents;
using Raven.Client.Documents.Session;
using Raven.Documentation.Samples.Orders;

namespace Raven.Documentation.Samples.ClientApi.Session.Attachments
{
    public class CopyMoveRename
    {
        private interface IFoo
        {
            #region copy_0
            void Copy(
                object sourceEntity,
                string sourceName,
                object destinationEntity,
                string destinationName);

            void Copy(
                string sourceDocumentId,
                string sourceName,
                string destinationDocumentId,
                string destinationName);
            #endregion

            #region rename_0
            void Rename(object entity, string name, string newName);

            void Rename(string documentId, string name, string newName);
            #endregion

            #region move_0
            void Move(object sourceEntity, string sourceName, object destinationEntity, string destinationName);

            void Move(string sourceDocumentId, string sourceName, string destinationDocumentId, string destinationName);
            #endregion
        }

        public async Task Sample()
        {
            using (var store = new DocumentStore())
            {
                using (var session = store.OpenSession())
                {
                    #region copy_1
                    var employee1 = session.Load<Employee>("employees/1-A");
                    var employee2 = session.Load<Employee>("employees/2-A");

                    session.Advanced.Attachments.Copy(employee1, "photo.jpg", employee2, "photo-copy.jpg");

                    session.SaveChanges();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region copy_2
                    var employee1 = await asyncSession.LoadAsync<Employee>("employees/1-A");
                    var employee2 = await asyncSession.LoadAsync<Employee>("employees/2-A");

                    asyncSession.Advanced.Attachments.Copy(employee1, "photo.jpg", employee2, "photo-copy.jpg");

                    await asyncSession.SaveChangesAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region rename_1
                    var employee = session.Load<Employee>("employees/1-A");

                    session.Advanced.Attachments.Rename(employee, "photo.jpg", "photo-new.jpg");

                    session.SaveChanges();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region rename_2
                    var employee = await asyncSession.LoadAsync<Employee>("employees/1-A");

                    asyncSession.Advanced.Attachments.Rename(employee, "photo.jpg", "photo-new.jpg");

                    await asyncSession.SaveChangesAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region move_1
                    var employee1 = session.Load<Employee>("employees/1-A");
                    var employee2 = session.Load<Employee>("employees/2-A");

                    session.Advanced.Attachments.Move(employee1, "photo.jpg", employee2, "photo.jpg");

                    session.SaveChanges();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region move_2
                    var employee1 = await asyncSession.LoadAsync<Employee>("employees/1-A");
                    var employee2 = await asyncSession.LoadAsync<Employee>("employees/2-A");

                    asyncSession.Advanced.Attachments.Move(employee1, "photo.jpg", employee2, "photo.jpg");

                    await asyncSession.SaveChangesAsync();
                    #endregion
                }
            }
        }
    }
}

File: Documentation/5.4/Samples/java/src/test/java/net/ravendb/DocumentExtensions/Attachments/CopyMoveRename.java
package net.ravendb.ClientApi.Session.attachments;

import net.ravendb.client.documents.DocumentStore;
import net.ravendb.client.documents.IDocumentStore;
import net.ravendb.client.documents.session.IDocumentSession;
import net.ravendb.northwind.Employee;

public class CopyMoveRename {

    private interface IFoo {
        //region copy_0
        void copy(Object sourceEntity, String sourceName,
                  Object destinationEntity, String destinationName);

        void copy(String sourceDocumentId, String sourceName,
                  String destinationDocumentId, String destinationName);
        //endregion

        //region rename_0
        void rename(Object entity, String name, String newName);

        void rename(String documentId, String name, String newName);
        //endregion

        //region move_0
        void move(Object sourceEntity, String sourceName,
                  Object destinationEntity, String destinationName);

        void move(String sourceDocumentId, String sourceName,
                  String destinationDocumentId, String destinationName);
        //endregion
    }

    public void sample() {
        try (IDocumentStore store = new DocumentStore()) {
            try (IDocumentSession session = store.openSession()) {
                //region copy_1
                Employee employee1 = session.load(Employee.class, "employees/1-A");
                Employee employee2 = session.load(Employee.class, "employees/2-A");

                session.advanced()
                    .attachments()
                    .copy(employee1, "photo.jpg", employee2, "photo-copy.jpg");

                session.saveChanges();
                //endregion
            }

            try (IDocumentSession session = store.openSession()) {
                //region rename_1
                Employee employee1 = session.load(Employee.class, "employees/1-A");

                session.advanced()
                    .attachments()
                    .rename(employee1, "photo.jpg", "photo-new.jpg");

                session.saveChanges();
                //endregion
            }

            try (IDocumentSession session = store.openSession()) {
                //region move_1

                Employee employee1 = session.load(Employee.class, "employees/1-A");
                Employee employee2 = session.load(Employee.class, "employees/2-A");

                session.advanced()
                    .attachments()
                    .copy(employee1, "photo.jpg", employee2, "photo.jpg");

                session.saveChanges();
                //endregion
            }
        }
    }
}
